### PR TITLE
[hlsl-out] clear `named_expressions` inserted by duplicated blocks

### LINKED ISSUE #1972: [hlsl-out] scoping issue with switch statement fallthrough
WGSL:

```
fn main() {
    var x = 0;
    switch (x) {
        case -1: {
            fallthrough;
        }
        default: {
            let var_2 = i32(11632);
        }
    }
}
```

HLSL:

```hlsl
void main()
{
    int x = 0;

    int _expr2 = x;
    switch(_expr2) {
        case -1: {
            {
            }
            {
                int var_2_ = int(11632);
            }
            break;
        }
        default: {
            int var_2_1 = var_2_;
            break;
        }
    }
}
```

`var_2_` is declared in the first case block but there is an invalid use in the `default` block. This looks like a recent regression since I think it was fine on an older commit of naga.

### PR DESCRIPTION
**Pull request by teoxoy**
fixes #1972

### CODE REVIEW

File: src/back/hlsl/writer.rs [hunk 1]
```diff
@@ -1876,6 +1877,27 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
...
+                    }
+                }
+                crate::Statement::Atomic { result: handle, .. }
+                | crate::Statement::Call {
+                    result: Some(handle),
+                    ..
+                } => {
+                    self.named_expressions.remove(&handle);
+                }
+                _ => {}
```
Comment (jimblandy) [inline src/back/hlsl/writer.rs:1896]: Doesn't this need to recurse into sub-statements?

**Comment by teoxoy** [comment]
Another way of doing this (which avoids the recursion) would be to use `indexmap::IndexMap` for `named_expressions` and use `.len()` + `.truncate(prev_len)`.

I'll push a commit that does this (which I can revert if we don't want to go this route).

**Comment by teoxoy** [comment]
Seems like we need https://github.com/bluss/indexmap/issues/215 for that to work...

**Comment by teoxoy** [comment]
I'll revert the last commit, adding `Arbitrary` support to `indexmap` also requires adding it to `hashbrown`. If we want to make it a newtype instead, it also gets really ugly since we need to expose all the methods of the inner `IndexMap` type.

**Comment by cuviper** [comment]
> adding `Arbitrary` support to `indexmap` also requires adding it to `hashbrown`.

I'm not sure why you think so... bluss/indexmap#246

**Comment by teoxoy** [comment]
@cuviper a PR, that's great! I was looking into deriving it (which would have required RawTable in hashbrown to also impl the trait).

**Comment by teoxoy** [comment]
@cuviper do you have an ETA on that PR landing and possibly a new release?

**Comment by cuviper** [comment]
Ah, deriving the fields directly would be *unlikely* to produce a coherent map. I just followed how each `Arbitrary` is implemented for a regular `HashMap`, with an iterator of `(K, V)` items to collect, or just `T` items for a set.

I went ahead and merged that on `master`. That's currently in a limbo 2.0.0 state though, which I should probably get around to publishing already, but I could also backport it to the 1.x series if you need that.

**Comment by teoxoy** [comment]
If it's not too much trouble, it would be greatly appreciated :)

**Comment by cuviper** [comment]
Done -- `indexmap 1.9.2`

**Comment by ErichDonGubler** [comment]
@jimblandy: Were you still going to review this for another round? I'm thinking of trying to tackle this.

**Comment by jimblandy** [comment]
Now that I actually understand what it's doing (from reviewing the switch PR), I was going to look it over again. Although, I'm pretty buried in #2075 right now.

**Review by ErichDonGubler — APPROVED** [review]
LGTM! Some non-blocking meta nits: 

* I sorta wish the justification for `indexmap` would land in the message of the commit introducing it, rather than just being GH discussion. That would make it easier for folks examining history. `git commit --fixup=amend…` is your friend!
* It's sorta odd to introduce a solution to the problem then immediately replace it in the same PR, but I don't see any issue with it.

File: tests/out/hlsl/control-flow.hlsl [hunk 2]
```diff
@@ -101,22 +101,36 @@ void main(uint3 global_id : SV_DispatchThreadID)
...
+            {
+            }
+            {
+                int i_1 = int(1);
+            }
+            break;
+        }
+        default: {
+            int i_2 = int(1);
+            break;
```
[the remark this answers is not in the document]
Comment (teoxoy) [reply tests/out/hlsl/control-flow.hlsl:194]: Yeah, we don't have any optimization passes. All the compilers naga targets have their own already.

**Comment by teoxoy** [comment]
Thanks for the review!

> It's sorta odd to introduce a solution to the problem then immediately replace it in the same PR, but I don't see any issue with it.

I left all commits in for now because I wanted some feedback on both approaches and was planning to squash everything when rebasing (before merging).

> I sorta wish the justification for `indexmap` would land in the message of the commit introducing it, rather than just being GH discussion. That would make it easier for folks examining history. `git commit --fixup=amend…` is your friend!

Will do.

**Comment by teoxoy** [comment]
Squashed and rebased. I had to remove the test since we removed support for fallthrough for the wgsl frontend in https://github.com/gfx-rs/naga/issues/2031.

**Review by ErichDonGubler — APPROVED** [review]
Still LGTM!

**Review by jimblandy — APPROVED** [review]
This looks great - much simpler than the original fix. It's a pity we can't test this.

**Pull request merged**